### ISSUE TITLE
underline, strickout and Arabic Render problem since bitmap combining

### ISSUE THREAD

**Issue by realfinder** [libass/libass#148]
vsfilter (and almost libass before bitmap)
![vsfilter](https://cloud.githubusercontent.com/assets/2512153/5061866/8a44668e-6dbb-11e4-95d9-5933f73b2905.png)

libass
![libass](https://cloud.githubusercontent.com/assets/2512153/5061864/869ea9ea-6dbb-11e4-8cd3-d8281fbe9d4e.png)

There are artifacts between letters

Script and font
http://www.mediafire.com/?39rbxfjgb8fmgmq

another one
http://www.mediafire.com/?dp70audha7rgpg7

more details and relation issue https://web.archive.org/web/20160120113701/https://code.google.com/p/libass/issues/detail?id=114#c6

note: this problem become worst after fix this https://github.com/libass/libass/issues/142

**Comment by grigorig** [comment]
I think we should ideally rasterize runs of text as a whole, not glyphs.

**Comment by astiob** [comment]
Yes! I’ve been of that opinion for ages, but it requires another fairly major change to the code, so there is no telling when we may get it done. Rasterizing whole runs will also help us considerably with VSFilter compatibility; even replicating VSFilter’s RTL behaviour will be trivial*.

\* Unless it actually resets to LTR every time, and it may well do; in which case a bit more work will be needed to get the expected behaviour for lines that use explicit RTL marks (overrides? I’m a bit hazy with these terms, sorry).

**Comment by anoosi** [comment]
is there plan to fix this issue?

**Comment by realfinder** [comment]
seems it not just the quality affected by this

[testing sub file.zip](https://github.com/libass/libass/files/5992583/testing.sub.file.zip) using avs assrender https://github.com/pinterf/assrender and avspmod
![image](https://user-images.githubusercontent.com/2512153/108142570-1656d100-70d7-11eb-93ff-fbcb7b1d9a37.png)

vsfilter (TextSub)
![image](https://user-images.githubusercontent.com/2512153/108142625-2d95be80-70d7-11eb-9db7-37055342902c.png)

both one by one
![test](https://user-images.githubusercontent.com/2512153/108144795-659f0080-70db-11eb-9912-42655c55345d.gif)

the bluring also is not same as vsfilter

also, as aside note, assrender don't load the font that inside the .ass file unlike vsfilter which did load it, it is problem in assrender or libass itself?

**Comment by astiob** [comment]
> also, as aside note, assrender don't load the font that inside the .ass file unlike vsfilter which did load it, it is problem in assrender or libass itself?

Most likely assrender, as libass does support reading embedded fonts. But libass did have some bugs in it in the past, so I suppose there’s a chance it’s using a bugged version of libass.

**Comment by TheOneric** [comment]
> But libass did have some bugs in it in the past

It still has bugs, or at least, embedded fonts only work with a certain callorder, which happens to match `mpv`'s (see #266).
Our `compare` tool, Aegisub and VLC cannot display embedded fonts.
